src/main/java/com/lightbend/akka/jpa/sample/UserRepository.java

```diff
@@ -0,0 +1,140 @@
+package com.lightbend.akka.jpa.sample;
+
+import akka.Done;
+import scala.concurrent.ExecutionContext;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.EntityTransaction;
+import javax.persistence.NoResultException;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+public class UserRepository {
+  // use withEntityManager
+  private final EntityManagerFactory entityManagerFactory;
+  private final ExecutionContext blockingDispatcher;
+
+  public UserRepository(EntityManagerFactory entityManagerFactory, ExecutionContext blockingDispatcher) {
+    this.entityManagerFactory = entityManagerFactory;
+    this.blockingDispatcher = blockingDispatcher;
+  }
+
+  public CompletionStage<Optional<User>> findById(long id) {
+    return withEntityManager(entityManager -> blockingFindById(entityManager, id));
+  }
+
+  // extracted because used from both findById and delete
+  private Optional<User> blockingFindById(EntityManager entityManager, long id) {
+    try {
+      CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+      CriteriaQuery<User> criteria = builder.createQuery(User.class);
+      Root<User> root = criteria.from(User.class);
+      criteria.select(root).where(builder.equal(root.get("id"), id));
+
+      User result = entityManager.createQuery(criteria).getSingleResult();
```

Comment:
Isn't there also a `EntityManager.detach` or something like that which tells JPA/Hibernate to explicitly stop monitoring the object?

Reply:
That is automatically done when closing the EntityManager (according to docs).

Reply:
That's also worth an explicit comment, because many JPA users might not be aware of it. The typical pattern (at least that I've seen) is closing the session at the very end of the request handling as in the [open session in view filter](https://vladmihalcea.com/the-open-session-in-view-anti-pattern/)... it may be widely considered a bad practice but I would bet that it's still very commonly used.

Reply:
Jup, that's a difference to highlight (and an advantage to call out, AFAIK excessively-long JPA sessions are still a common source of tricky performance problems and this approach is a 'solution' that should still look familiar to people with a JPA background)